common/dynamicconfig/memory_client.go

```diff
@@ -90,43 +93,65 @@ func (d *MemoryClient) OverrideSetting(setting GenericSetting, value any) (clean
 
 func (d *MemoryClient) OverrideValue(key Key, value any) (cleanup func()) {
 	d.lock.Lock()
-	defer d.lock.Unlock()
 
 	var idx atomic.Int64
 	idx.Store(int64(len(d.overrides)))
 
 	d.overrides = append(d.overrides, kvpair{valid: true, key: key, value: value})
-
 	newValue := d.getValueLocked(key)
 	changed := map[Key][]ConstrainedValue{key: newValue}
-	for _, update := range d.subscriptions {
-		update(changed)
+
+	// Copy out subscriptions to call after we release the lock
+	subscribers := make([]ClientUpdateFunc, 0, len(d.subscriptions))
+	for _, sub := range d.subscriptions {
+		subscribers = append(subscribers, sub)
+	}
+
+	d.lock.Unlock()
+
+	// Notify subscribers *outside* the lock to avoid deadlock
+	for _, updateFunc := range subscribers {
+		updateFunc(changed)
 	}
 
 	return func() {
-		// only do this once
+		// Cleanup only once
 		if removeIdx := int(idx.Swap(-1)); removeIdx >= 0 {
+			// We still need to remove the override, but again
+			// release the lock before notifying subscribers.
 			d.remove(removeIdx)
 		}
 	}
 }
 
 func (d *MemoryClient) remove(idx int) {
 	d.lock.Lock()
-	defer d.lock.Unlock()
+
+	if idx < 0 || idx >= len(d.overrides) || !d.overrides[idx].valid {
+		d.lock.Unlock()
+		return
+	}
 
 	key := d.overrides[idx].key
-	// mark this pair deleted
-	d.overrides[idx] = kvpair{}
+	d.overrides[idx].valid = false
```

Comment:
I don't like this change, I deliberately reset the whole struct to drop the reference to the old value. This keeps the reference alive unnecessarily.